### PR TITLE
Optimize string operations throughout the codebase

### PR DESCRIPTION
**Pull request by MewPurPur**
A more manageable version of #82412 that does more types of optimizations, but only applies them to user-exposed classes (so the total volume of them is a lot less). No benchmarking done.

For reference, most of them are like:

- `s.substr(0, x)` -> `s.left(x)`
- `s.substr(0, s.length() - x)` -> `s.left(-x)`
- `s.substr(x, s.length())` -> `s.right(-x)`
- `s.substr(x, s.length() - x)` -> `s.right(-x)`

I think all of these are equivalent 100% of the time, but they are ~15% faster (benchmarking from my PR that optimized left and right).

### CODE REVIEW

File: core/io/file_access_compressed.cpp [hunk 1]
```diff
@@ -33,8 +33,7 @@
 #include "core/string/print_string.h"
 
 void FileAccessCompressed::configure(const String &p_magic, Compression::Mode p_mode, uint32_t p_block_size) {
-	magic = p_magic.ascii().get_data();
-	magic = (magic + "    ").substr(0, 4);
+	magic = p_magic.ascii().get_data().rpad(4, " ");
```
Comment (AThousandShips) [inline core/io/file_access_compressed.cpp:36]: This calls `rpad` on `comst char *`

File: core/io/http_client.cpp [hunk 2]
```diff
@@ -117,8 +117,8 @@ Dictionary HTTPClient::_get_response_headers_as_dictionary() {
 		if (sp == -1) {
 			continue;
 		}
-		String key = s.substr(0, sp).strip_edges();
-		String value = s.substr(sp + 1, s.length()).strip_edges();
+		String key = s.left(sp).strip_edges();
+		String value = s.right(-(sp + 1)).strip_edges();
```
Comment (AThousandShips) [inline core/io/http_client.cpp:121]: This feels very hard to read, I don't think it's an improvement readability wise

File: core/string/translation_po.cpp [hunk 3]
```diff
@@ -172,10 +172,10 @@ void TranslationPO::_cache_plural_tests(const String &p_plural_rule) {
 		return;
 	}
 
-	String equi_test = p_plural_rule.substr(0, first_ques_mark).strip_edges();
+	String equi_test = p_plural_rule.left(first_ques_mark).strip_edges();
 	equi_tests.push_back(equi_test);
 
-	String after_colon = p_plural_rule.substr(p_plural_rule.find(":") + 1, p_plural_rule.length());
+	String after_colon = p_plural_rule.right(-(p_plural_rule.find(":") + 1));
```
Comment (AThousandShips) [inline core/string/translation_po.cpp:178]: Ditto

File: core/string/ustring.cpp [hunk 4]
```diff
@@ -231,20 +231,20 @@ Error String::parse_url(String &r_scheme, String &r_host, int &r_port, String &r
 	int pos = base.find("://");
 	// Scheme
 	if (pos != -1) {
-		r_scheme = base.substr(0, pos + 3).to_lower();
-		base = base.substr(pos + 3, base.length() - pos - 3);
+		r_scheme = base.left(pos + 3).to_lower();
+		base = base.right(-pos - 3);
```
Comment (AThousandShips) [inline core/string/ustring.cpp:235]: Ditto

File: core/string/ustring.cpp [hunk 5]
```diff
@@ -231,20 +231,20 @@ Error String::parse_url(String &r_scheme, String &r_host, int &r_port, String &r
...
+		r_path = base.right(-pos);
+		base = base.left(pos);
 	}
 	// Host
 	pos = base.find("@");
 	if (pos != -1) {
-		// Strip credentials
-		base = base.substr(pos + 1, base.length() - pos - 1);
+		// Strip credentials.
+		base = base.right(-pos - 1);
```
Comment (AThousandShips) [inline core/string/ustring.cpp:247]: Ditto

File: core/string/ustring.cpp [hunk 6]
```diff
@@ -253,7 +253,7 @@ Error String::parse_url(String &r_scheme, String &r_host, int &r_port, String &r
 			return ERR_INVALID_PARAMETER;
 		}
 		r_host = base.substr(1, pos - 1);
-		base = base.substr(pos + 1, base.length() - pos - 1);
+		base = base.right(-pos - 1);
```
Comment (AThousandShips) [inline core/string/ustring.cpp:256]: Ditto

File: scene/gui/file_dialog.cpp [hunk 7]
```diff
@@ -800,8 +800,8 @@ void FileDialog::set_current_path(const String &p_path) {
 	if (pos == -1) {
 		set_current_file(p_path);
 	} else {
-		String path_dir = p_path.substr(0, pos);
-		String path_file = p_path.substr(pos + 1, p_path.length());
+		String path_dir = p_path.left(pos);
+		String path_file = p_path.right(-(pos + 1));
```
Comment (AThousandShips) [inline scene/gui/file_dialog.cpp:804]: Ditto

File: scene/gui/rich_text_label.cpp [hunk 8]
```diff
@@ -4059,7 +4059,7 @@ void RichTextLabel::append_text(const String &p_bbcode) {
 				const String &expr = split_tag_block[i];
 				int value_pos = expr.find("=");
 				if (value_pos > -1) {
-					bbcode_options[expr.substr(0, value_pos)] = expr.substr(value_pos + 1).unquote();
+					bbcode_options[expr.left(value_pos)] = expr.right(-value_pos - 1).unquote();
```
Comment (AThousandShips) [inline scene/gui/rich_text_label.cpp:4062]: Ditto

File: scene/gui/rich_text_label.cpp [hunk 9]
```diff
@@ -4638,8 +4638,8 @@ void RichTextLabel::append_text(const String &p_bbcode) {
 					if (sep == -1) {
 						width = bbcode_value.to_int();
 					} else {
-						width = bbcode_value.substr(0, sep).to_int();
-						height = bbcode_value.substr(sep + 1).to_int();
+						width = bbcode_value.left(sep).to_int();
+						height = bbcode_value.right(-sep - 1).to_int();
```
Comment (AThousandShips) [inline scene/gui/rich_text_label.cpp:4642]: Ditto

File: scene/gui/rich_text_label.cpp [hunk 10]
```diff
@@ -4675,28 +4675,28 @@ void RichTextLabel::append_text(const String &p_bbcode) {
...
-			int fnt_size = tag.substr(10, tag.length()).to_int();
+			int fnt_size = tag.right(-10).to_int();
 			push_font_size(fnt_size);
 			pos = brk_end + 1;
 			tag_stack.push_front("font_size");
 
 		} else if (tag.begins_with("opentype_features=") || tag.begins_with("otf=")) {
 			int value_pos = tag.find("=");
-			String fnt_ftr = tag.substr(value_pos + 1);
+			String fnt_ftr = tag.right(-value_pos - 1);
```
Comment (AThousandShips) [inline scene/gui/rich_text_label.cpp:4699]: Ditto

**Comment by AThousandShips** [comment]
It appears the changes breaks some code, see the broken doc generation

File: core/io/http_client.cpp (hunk 2 above)
Comment (MewPurPur) [reply core/io/http_client.cpp:121]: I tend to agree. I think `s.substr(pos)` is the most readable, so I'm checking now if I can optimize substr(pos) by just calling right(-pos) if a second argument isn't passed.

**Comment by AThousandShips** [comment]
Unit tests show their worth, let's see what's broken with your implementation of the string methods, will do some looking into it tomorrow if you haven't solved 🙂

**Comment by MewPurPur** [comment]
I can't figure it out :disappointed:

**Comment by kleonc** [comment]
> I think all of these are equivalent 100% of the time


These are not the same for `x < 0`:
>* `s.substr(0, x)` -> `s.left(x)` 
>* `s.substr(x, s.length())` -> `s.right(-x)`

These are not the same for `x <= 0`:
>* `s.substr(0, s.length() - x)` -> `s.left(-x)`
>* `s.substr(x, s.length() - x)` -> `s.right(-x)`

---

> but they are ~15% faster (benchmarking from my PR that optimized left and right).

The mentioned PR: #80824 (please link next time you refer to some other PR).
Its description is a little unclear to me but I understand it as the benchmarked speedup is for old `left/right` vs new `left/right`. In such case it doesn't say anything about the performance of `left/right` vs `substr`, and this would need to be benchmarked.
But I'd rather focus on trying to optimize `substr` in the first place, e.g. are the `length()` calls optimized out by the compiler and done once? If not then these should cached/ called once manually (and used in check instead of `is_empty()` etc.).
https://github.com/godotengine/godot/blob/2d0ee20ff30461b6b10f6fdfba87511a0ebc6642/core/string/ustring.cpp#L2975-L2995
Same for other methods (like `left/right`). Note it's already like that e.g. in `find`:
https://github.com/godotengine/godot/blob/2d0ee20ff30461b6b10f6fdfba87511a0ebc6642/core/string/ustring.cpp#L3002-L3004

---

Regarding code simplification these can be done for sure:
* `s.substr(x, s.length())` -> `s.substr(x)`
* `s.substr(x, s.length() - x)` -> `s.substr(x)`

I find these better readability-wise, plus it gets rid of redundant `s.length()` calls on the caller side (`length()` is called in `substr` anyway).

**Comment by AThousandShips** [comment]
I'd say that unless things are in a very performance critical area readability trumps performance, I'm in favor of the obvious ones like `substr(x, length)`, but generally to stay on the safe side as `left/right` can be confusing at times, and need to ensure properly the correct conversions

I think this should be done in two PRs, one to improve the performance, and one to replace relevant cases

**Comment by MewPurPur** [comment]
Hmpf, so my old PR only did ones with `left()`, I thought `right()` would be about as easy but it's not.

I know that `substr()` is confusing for a lot of people in a way similar to ternaries, where you constantly have to check if the second argument was the final position, or the substring's length (i.e. does `"ABCDEF".substr(1, 3)` give BCD or BC?) so I always prefer the dedicated methods when the substring isn't in the middle of the string. But while `left()` tends to always be readable, `right()` just sucks for readability when you flip the sign IMO.

I think I'll abandon the idea because of that.

> Its description is a little unclear to me but I understand it as the benchmarked speedup is for old left/right vs new left/right. In such case it doesn't say anything about the performance of left/right vs substr, and this would need to be benchmarked.

The old left/right were calling substr() directly, so it seems fair to me. Unless there is a notable penalty for going through an extra function that I don't know of.

**Comment by AThousandShips** [comment]
I'd say to remove the potentially ambiguous cases and see if that resolves the CI, that being the `right` cases with negative indices at least

The `substr` cases with trivial length should be good IMO, as are the trivial `left` cases, i.e. `substr(0, x)`, and maybe do `right` in a future PR with more discussion and evaluation

**Comment by MewPurPur** [comment]
What the heck

**Comment by MewPurPur** [comment]
I'll leave this idea at peace for now.

**Pull request closed without merging**